Release/1.0.0

istreeter opened the pull request:
Add from URI constructors to emitters (#116)
UUIDProviders for Id and Sync (#112)
Merge the single and batch id-emitters (#142)
Refactor thread usage for the id-emitters (#143)
Add ADT for retry policy on failed http requests (#144)
Add ADT for event queue size policy (#145)
Add a Buffer class (#146)
Add ability to flush Emitter cache (#119)
Use a Action ADT to signal different messages to the Emitter (#147)
Allow subject to be set for specific events (#148)
Tracker method to set metadata after initialisation (#149)
Add a http4s emitter (#115)
Reduce MetadataSpec's flakiness (#110)

paulboocock reviewed (approved):
LGTM. 

This is a nice improvement, particularly for those who are less "scala-y" (like me!). I'd love to see a review someone who has integrated this recently since it's going to be v1 (@andreamarcolin?).

Last note: Can the README be updated from Snowplow Early Release to Snowplow Maintained?

```
[![maintained]][tracker-classificiation]

[tracker-classificiation]: https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/tracker-maintenance-classification/
[maintained]: https://img.shields.io/static/v1?style=flat&label=Snowplow&message=Maintained&color=9e62dd&labelColor=9ba0aa&logo=data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABAAAAAQCAMAAAAoLQ9TAAAAeFBMVEVMaXGXANeYANeXANZbAJmXANeUANSQAM+XANeMAMpaAJhZAJeZANiXANaXANaOAM2WANVnAKWXANZ9ALtmAKVaAJmXANZaAJlXAJZdAJxaAJlZAJdbAJlbAJmQAM+UANKZANhhAJ+EAL+BAL9oAKZnAKVjAKF1ALNBd8J1AAAAKHRSTlMAa1hWXyteBTQJIEwRgUh2JjJon21wcBgNfmc+JlOBQjwezWF2l5dXzkW3/wAAAHpJREFUeNokhQOCA1EAxTL85hi7dXv/E5YPCYBq5DeN4pcqV1XbtW/xTVMIMAZE0cBHEaZhBmIQwCFofeprPUHqjmD/+7peztd62dWQRkvrQayXkn01f/gWp2CrxfjY7rcZ5V7DEMDQgmEozFpZqLUYDsNwOqbnMLwPAJEwCopZxKttAAAAAElFTkSuQmCC 
```

chuwy commented on modules/id-emitter/src/main/scala/com/snowplowanalytics/snowplow/scalatracker/emitters/id/AsyncEmitter.scala:
```diff
@@ -27,55 +30,138 @@ object AsyncEmitter {
    * Backed by `java.util.concurrent.LinkedBlockingQueue`, which has
    * capacity of `Int.MaxValue` will block thread when buffer reach capacity
    *
+   * This emitter sends requests asynchronously from the tracker's meain thread of execution,
```
```suggestion
   * This emitter sends requests asynchronously from the tracker's main thread of execution,
```

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Buffer.scala:
```diff
@@ -0,0 +1,106 @@
+  import Buffer._
+
+  /* Resets the internal state to an empty buffer, i.e. after sending the pending batch in a post request
+   */
```
Just noticed that ScalaDoc formatting here is slightly off compared to other modules.

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Buffer.scala:
```diff
@@ -0,0 +1,106 @@
+
+  private case class BufferImpl(toList: List[Payload], count: Int, bytes: Int, config: BufferConfig) extends Buffer {
+
+    override def reset: Buffer = BufferImpl(Nil, 0, 0, config)
```
```suggestion
    override def reset: Buffer = Buffer(config)
```

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Buffer.scala:
```diff
@@ -0,0 +1,106 @@
+        case BufferConfig.NoBuffering =>
+          toList.nonEmpty
+        case BufferConfig.EventsCardinality(max) =>
+          count >= max && count > 0
```
Shouldn't we instead prohibit negative values in `BufferConfig`?

istreeter replied:
I suggest I remove the `count > 0` clause - It is redundant code because nothing bad happens if if `max` is zero or negative.

Regarding prohibiting zero/negative values - this can be done with either
- assert/require statements in class body.  I really don't like this method.
- refinement types, e.g. https://github.com/fthomas/refined.  Probably overkill here.

chuwy replied:
I had refinement types on mind, but actually agree it's a huge overkill, especially if we don't use them in other places.

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Payload.scala:
```diff
@@ -14,65 +14,76 @@ package com.snowplowanalytics.snowplow.scalatracker
-    val filtered = dict.filter { case (key, value) => key != null && value != null && !key.isEmpty && !value.isEmpty }
-    Payload(nvPairs ++ filtered)
-  }
+  private[scalatracker] implicit class PayloadSyntax(payload: Payload) {
```
```suggestion
  private[scalatracker] implicit class PayloadSyntax(payload: Payload) extends AnyVal {
```

In fact, `AnyVal` is so bad in doing what it promises (avoiding unnecessary boxing) that the only place where it's 100% safe is `Ops` classes.

istreeter replied:
This is not immediately possible, because of this compiler error:
> value class may not wrap another user-defined value class

It's because Payload already extends AnyVal.

I see 3 options:
1. leave it as it is.
2. Remove the `extends AnyVal` from Payload and put it instead on PayloadSyntax
3. Remove the PayloadSyntax class entirely and add these methods directly as properties of the Payload class.

Hmmm... option 3 is probably the best suggestion.  wdyt @chuwy ?

chuwy replied:
Ah, ok sorry - I didn't realise `Payload` is `AnyVal` as well. My gut feeling tells me we'll have less boxings if we use `AnyVal` for this class rather than for `Payload`, but 1) I have absolutely no proofs; 2) Even in that case it will be just a micro-optimisation.

Marking as solved.

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Tracker.scala:
```diff
@@ -59,22 +59,22 @@ final case class Tracker[F[_]: Monad: Clock: UUIDProvider](emitters: NonEmptyLis
-    emitters.traverse(e => e.send(event)).map(x => ())
-  }
+  private def send(payload: Payload): F[Unit] =
+    emitters.traverse(e => e.send(payload)).void
```
```suggestion
    emitters.traverse_(e => e.send(payload))
```

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/UUIDProvider.scala:
```diff
@@ -0,0 +1,45 @@
+ * This trait is needed here because `Tracker` is referentially transparent, and we leave it up for the emitters
+ * to decide whether `F` will be referentially transparent also.
+ */
+trait UUIDProvider[F[_]] {
```
I saw your comment in https://github.com/snowplow/snowplow-scala-tracker/issues/112. I think the perfect solution here would be to use some `Random` capability library, i.e. randomness is not impure as IO. But I don't know of any such libraries in cats ecosystem. This is the closest one: https://github.com/NICTA/rng, but super scalaz-legacy.

chuwy commented on modules/http4s-emitter/src/main/scala/com.snowplowanalytics.snowplow.scalatracker.emitters.http4s/Http4sEmitter.scala:
```diff
@@ -0,0 +1,153 @@
+    policy match {
+      case EventQueuePolicy.UnboundedQueue      => Queue.unbounded
+      case EventQueuePolicy.BlockWhenFull(max)  => Queue.bounded(max)
+      case EventQueuePolicy.IgnoreWhenFull(max) => Queue.bounded(max)
```
Wouldn't it be `circularBuffer`?

chuwy replied:
If we're happy with `bounded` everywhere it can be:

```scala
  private def queueForPolicy[F[_]: Concurrent](policy: EventQueuePolicy): F[Queue[F, Action]] =
    policy.tryLimit match {
      case None      => Queue.unbounded
      case Some(max) => Queue.bounded(max)
    }
```

chuwy replied:
Also it seems we don't actually use it in `Http4sEmitter`?

istreeter replied:
I did not use circularBuffer for two reasons:
- If the queue is full then it is not obvious whether to drop the *newest* event or the *oldest* event. One can make a strong case for either.
- CircularBuffer is a fs2 class, and there is no equivalent class to be used in the id-emitters.  I wanted the EventQueuePolicy to have the same semantic meaning for both sets of emitters.

andreamarcolin commented:
Hey @istreeter! As I said to @paulboocock, I would love to try this out before the official release (especially the new http4s emitter), but I don’t think I will be able to review the PR. It’s huge and I don’t know enough about trackers to provide meaningful insights. I will limit myself to just reviewing its integration as a library user, even if I'm sure you already did that ;)

chuwy commented on project/Dependencies.scala:
```diff
@@ -21,6 +21,7 @@ object Dependencies {
     val circe       = "0.13.0"
     val circeOptics = "0.13.0"
     val catsEffect  = "2.2.0"
+    val http4s      = "0.21.9"
```
0.21.13 is available.

istreeter commented:
Thanks @andreamarcolin it would be really helpful if you try out integrating it as a user.

The http4s-emitter is very "catsy" (new word!) and I think it should integrate nicely.  I am more worried about the id-emitters, because they should be targeted at users who are not using the cats-effect ecosystem, e.g. akka, spark or play frameworks.  It would be interesting if you could take that persona and then comment on their ease of setup.

istreeter commented:
A few minor changes since @paulboocock last reviewed:
- Made `flushBuffer` into a field on the base Emitter trait, rather than a emitter-specific implementation.
- Therefore made `flushEmitters` a new method on the Tracker.
- Changed readme to 'snowplow-maintained'
- Removed the commit for issue #149.  The tracker is a case class, so it automatically has a `copy` method, and therefore does not make sense to add an extra `setMetadata` method.  i.e. can just do `tracker.copy(metadata = Some(metadata))` instead.
- Changed the name of the [id-specific tracker](https://github.com/snowplow/snowplow-scala-tracker/blob/release/1.0.0/modules/id-emitter/src/main/scala/com/snowplowanalytics/snowplow/scalatracker/emitters/id/IdTracker.scala) from Tracker to IdTracker.  I am struggling a bit with the class - I feel it needs to exist so that users of the id-emitters don't need to implement their own `Clock[Id]` just to initialize a tracker.  I'm struggling on where to place it in the package, but I think calling it IdTracker makes sense.

andreamarcolin commented:
Hey @istreeter, I apologize for the long wait. I was about to try to upgrade snowplow-scala-tracker-emitter-id_2.13 to 1.0.0-M1 but apparently it didn't made it to maven central, not sure why. 
We are going to use the brand new snowplow-scala-tracker-emitter-http4s_2.13 (we are on cats/cats-effect/http4s), but my plan was to first bump the id-emitter to 1.0.0-M1, see if everything looks good to give you some feedback, and then switch to the new one. Unfortunately it got pusblished only for 2.12 and I have no means of easily going back to 2.12 in the project I am using as a test. 
Do you think we can get the 2.13 version on Maven Central?

paulboocock commented:
> Do you think we can get the 2.13 version on Maven Central?

_sigh_
The Maven Central sync from bintray is so flakey. I thought I'd sorted them out (I noticed this during my review so I manually synced them). I'll give them another poke now, one can hope it'll be available tomorrow.

istreeter commented:
@andreamarcolin your plan sounds very helpful; i.e. bump your id-emitter and then afterwards switch to the http4s-emitter.

Please can you use the 1.0.0-M2 asset?  There's not much difference from M1, but there is one important bugfix in M2.

For plg we are pulling in the tracker from [bintray](https://bintray.com/snowplow/snowplow-maven/snowplow-scala-tracker-emitter-id) by adding this to our sbt build:
```
  lazy val resolverSettings = Seq[Setting[_]](
    resolvers ++= Seq(
      "Sonatype OSS Snapshots".at("https://oss.sonatype.org/content/repositories/snapshots/"),
      "Snowplow Bintray" at "https://snowplow.bintray.com/snowplow-maven"
    )
  )
```

Alternatively - let's hope Paul's intervention fixes the problem in maven central by tomorrow.

paulboocock commented:
It appears that the issue here is that JCenter isn't syncing the new Scala 2.13 versions of the packages from our bintray. Probably due to [this issue](https://stackoverflow.com/questions/40963068/jcenter-and-bintray-not-in-sync). `snowplow-scala-tracker-emiiter-http4s` is working as this has only just been included in JCenter and seems to have been set up correctly for 2.12 and 2.13.

Older packages seem to not be syncing any `_2.13` packages to JCenter, so currently we're unable to publish `snowplow-scala-tracker-emitter-id`, `snowplow-scala-tracker-core` and `snowplow-scala-tracker-metadata` to JCenter (and therefore unable to publish to Maven Central, as the Maven Central sync works from JCenter). 

I've opened a support ticket with Bintray, I'll let you know when it is resolved and I've synced the latest Milestone releases to JCenter and Maven Central. Hopefully when this is fixed, the publishing should work as expected too.

@andreamarcolin In the meantime, you can do what @istreeter suggests above as all the assets are available correctly via the Snowplow Maven on Bintray.

andreamarcolin reviewed (commented):
Hey @istreeter! Had the chance to quickly bump our project to 1.0.0-M2, and it was still compiling, which was welcome news :)

So I read about the possibility of not manually providing instances for Clock[Id] and UUIDProvider[Id], but I had to dig into this PR to find the IdTracker smart constructor, which was somehow unexpectedly inside the `emmitters.id` package (probably just because I am new to the library). 
Not sure if docs where updated (didn't check), but I think this is the first thing you want to know when integrating scala-tracker into your code.
Nonetheless, swapping the Tracker constructor for IdTracker and removing the manually provided typeclass instance was a nice improvement, and I think it lowers the barrier a lot for developers who are not embracing FP and typeclasses in general to be able to setup the tracker without knowing about them.

Another improvement I was expecting but couldn't find is a smart constructor for `Emitter`s (SyncEmitter in my case) accepting a URI. I can see that internally the `createAndStart` one uses EndpointParams, which in turn provides a `fromUri` smart constructor. It would be good if we could provide this functionality as described in https://github.com/snowplow/snowplow-scala-tracker/issues/116.

Overall, I had no major problem upgrading the library anyway, so... great job!
Next thing on the list: swap out the SyncEmitter for the new Http4s emitter :)

istreeter commented:
Thanks for the review @andreamarcolin 

> Not sure if docs where updated

Not yet!  I plan to do this today.

> Another improvement I was expecting but couldn't find is a smart constructor for Emitters (SyncEmitter in my case) accepting a URI.

You are correct - I should change the `createAndStart` methods to accept `EndpointParams` instead of accepting host, port, and https.  Then you could use them something like this:

```scala
EndpointParams.fromURI(new URI("http://example.com")) match {
  case Right(endpointParams) => SyncEmitter.createAndStart(endpointParams)
  case Left(error) => throw new RuntimeException(s"cannot create emitter $error")
}
```

Do you agree that change would be good enough to satisfy #116?  Or should I go a step further so the developer can do this....

```scala
val emitter = SyncEmitter.createAndStart(new URI("http://example.com")) // What happens if uri is invalid?
```

My preference is to not go that extra step.  At least we have offered an easy way for the developer to create a EndpointParams from a URI.

> So I read about the possibility of not manually providing instances for Clock[Id] and UUIDProvider[Id]........

I really struggled with this!  I cannot decide what helps developers most, or what fits their expectations.

I put IdTracker in the id-emitters module because it only exists to support those emitters, but I could move it to core.  Another option is to create a idimplicits module, so the developer would do something like this (instead of importing IdTracker):

```scala
import com.snowplowanalytics.snowplow.scalatracker.idimplicits._
val idEmitter = SyncEmitter.createAndStart(endpointParams)
val tracker = Tracker(idEmitter, "namespace", "appid")
```

wdyt?

istreeter commented:
Updates 2020-12-02:

- Removed the IdTracker and added a idimplicits module in its place.  I decided the idimplicits pattern is more inline with how other libraries provide implicit requirements
- The Sync and Async emitters now take EndpointParams case class, instead of host, port, https.  This means developers can use the `EndpointParams.fromURI` method, which better satisfies issue #116
- Renamed the SyncEmitters constructor from `createAndStart` to just `apply`.  It starts nothing (!!) so the new name is more appropriate
- Added documentation to the [snowplow docs site](https://docs.snowplowanalytics.com/docs/collecting-data/collecting-from-own-applications/scala-tracker/scala-tracker-1-0-0/)

andreamarcolin commented:
> Do you agree that change would be good enough to satisfy #116?

Absolutely agree. I don't think I would ever suggest to throw an exception there, not even in OOP Scala. I think returning an Either is the best option there, even in non-FP Scala. The user can always decide if they want to throw for the left case, while on the other way around (throwing), you force a user to look inside for what to catch, and eventually he would even have to wrap that handling in a Either, which wouldn't be pleasant to see in the code imho.

> Removed the IdTracker and added a idimplicits module in its place. I decided the idimplicits pattern is more inline with how other libraries provide implicit requirements

This makes sense to me, and I like it more than the previous approach. An implicits package to import is something widely used in FP Scala, but Scala itself uses that pattern (think scala.concurrent.ExecutionContext.Implicits), so I think it's fine for OOP/procedural styles too.

> Renamed the SyncEmitters constructor from createAndStart to just apply. It starts nothing (!!) so the new name is more appropriate

Ahh this one! Definitely makes sense. Makes it a lot less scary to me also! I am always worried about things which say they start other things :D

> Added documentation to the snowplow docs site

I read through it, looks awesome!

Excellent job @istreeter, massive step forward for the Scala Tracker. LGTM now! Looking forward to the final release :)

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Emitter.scala:
```diff
@@ -38,20 +26,288 @@ trait UUIDProvider[F[_]] {
+    /** Configures the emitter to buffer events and send batched payloads comprising a fixed number of events
+     *  @param size The number of events after which the emitter sends the buffered events to the collector
+     */
+    case class EventsCardinality(size: Int) extends BufferConfig
```
Wondring if we can add some kind of `OR`, i.e. "send request when events cardinality reached X OR when payload size reached Y".

chuwy replied:
Did you actually do it?! I was thinking to add a ticket for it.

istreeter replied:
😆 It was a fairly easy change, so I just did it.

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Emitter.scala:
```diff
@@ -38,20 +26,288 @@ trait UUIDProvider[F[_]] {
+  sealed trait Request extends Product with Serializable {
+
+    /** Attempt to send */
+    def attempt: Int
```
I think we need to clarify that it's _next attempt_, not _amount of times we've tried to submit it_, i.e. it's 1-based, not 0-based. Have no idea how to phrase it myself.

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Emitter.scala:
```diff
@@ -38,20 +26,288 @@ trait UUIDProvider[F[_]] {
+     *
+     * @param max The maximum number of http requests for a batch of events is dropped.
+     */
+    case class MaxAttempts(max: Int) extends RetryPolicy
```
Noticed that in rest of ADTs you have `final` modifier.

chuwy commented on modules/core/src/main/scala/com.snowplowanalytics.snowplow/scalatracker/Tracker.scala:
```diff
@@ -36,7 +36,7 @@ import scala.concurrent.duration._
 final case class Tracker[F[_]: Monad: Clock: UUIDProvider](emitters: NonEmptyList[Emitter[F]],
```
1. I don't think it's a proper case class, e.g. we cannot provide meaningful `toString` and `equals`.
2. In more tagless-final-ish approach we have type class constraints on functions, not on data types

Both are minor though and I think addressing 2nd point might require a big API change.

istreeter replied:
I changed to a regular class, not case class.

I chose to keep the class constraints on the class, rather than the functions.  The other way around would require the developer to have a UUIDProvider and Clock in scope every time they call `tracker.trackSomething(....)`.  And for the id-emitters I am not fully satisfied with how we provide a UUIDProvider[Id] and a Clock[Id] - so I would prefer it if these are required only for the tracker constructors, rather than throughout the developer's entire codebase.

Good suggestions though!!

chuwy commented on modules/http4s-emitter/src/main/scala/com.snowplowanalytics.snowplow.scalatracker.emitters.http4s/Http4sEmitter.scala:
```diff
@@ -0,0 +1,146 @@
+                                                                result: Result): F[Unit] =
+    callback match {
+      case Some(cb) =>
+        cb(collector, request, result).handleError(_ => ())
```
I think we should force users to handle errors in callback, otherwise I can imagine a typical handling scenario where all unsent payloads are written to the disk and the disk runs out of space.

istreeter replied:
I think what you are saying is that if `cb` raises an error, then we should ignore it here, and let it bubble up into the surrounding thread of execution.

This might not have the effect you originally think though.... the fiber that invokes the callback is completely isolated from the surrounding program's main fiber of execution.  So this fiber would die, but the program's main fiber would be unaware.  It is hard (not impossible but hard) to set things up in such a way that an exception in `ct` causes the program to exit.

The same is true of the AsyncEmitter in the id-emitters module.

I think I prefer to catch the exception, because it is easier to reason about the resulting behaviour.

chuwy replied:
My reasoning was more about encouraging users to handle all errors rahter than closing an app (although that would be good as well). Because in case of `handleError(_ => ())` if something bad happens they won't even see it in logs, but ideally they should be encouraged to do something like sentry tracking.

chuwy commented on modules/http4s-emitter/src/main/scala/com.snowplowanalytics.snowplow.scalatracker.emitters.http4s/Http4sEmitter.scala:
```diff
@@ -0,0 +1,146 @@
+                                             retryPolicy: RetryPolicy,
+                                             rng: Random,
+                                             callback: Option[Callback[F]]): Request => F[Unit] =
+    Sync[F].tailRecM(_) { request =>
```
I honestly never seen this pattern (using `tailRecM` for retries), but looks interesting.

chuwy commented on modules/http4s-emitter/src/main/scala/com.snowplowanalytics.snowplow.scalatracker.emitters.http4s/Http4sEmitter.scala:
```diff
@@ -0,0 +1,146 @@
+    }
+
+  private def shutdown[F[_]: Monad](fiber: Fiber[F, Unit], queue: Enqueue[F, Action]): F[Unit] =
+    queue.enqueue1(Action.Terminate) *> fiber.join
```
Do we want to have a timeout?

chuwy commented on modules/http4s-emitter/src/test/scala/com/snowplowanalytics/snowplow/scalatracker/emitters/http4s/Http4sEmitterSpec.scala:
```diff
@@ -0,0 +1,160 @@
+
+class Http4sEmitterSpec extends Specification {
+
+  implicit val cs: ContextShift[IO] = IO.contextShift(ExecutionContext.global)
```
You can have a look at https://github.com/djspiewak/cats-effect-testing/#specs2

chuwy reviewed (approved):
That's really-really great work, @istreeter! Have left some comments, but nothing major really.
